lisa/tests/hotplug/rollback.py

```diff
@@ -0,0 +1,195 @@
+
+from devlib.module.hotplug import HotplugModule
+from devlib.exception import TargetStableError
+
+from lisa.tests.base import ResultBundle, TestBundle
+from lisa.trace import FtraceCollector, Trace
+from lisa.utils import ArtifactPath
+
+
+class HotplugRollback(TestBundle):
+
+    TRACE_PATH = 'trace.dat'
+
+    @classmethod
+    def _online(cls, target, cpu, online, verify=True):
+        try:
+            if online:
+                target.hotplug.online(cpu)
+            else:
+                target.hotplug.offline(cpu)
+        except Exception as e:
+            if verify:
+                raise e
+
+    @classmethod
+    def _state_can_fail(cls, target, cpu, state, up):
+        """
+        There are no way of probing the kernel for a list of hotplug states that
+        can fail and for which we can test the rollback. We need therefore to
+        try:
+          - If we can set the state in the kernel 'fail' interface.
+          - If the hotplug is reset actually failing (some states can fail only
+            when going up or down)
+        """
+        try:
+            target.hotplug.fail(cpu, state)
+            try:
+                cls._online(target, cpu, up)
+                target.hotplug.fail(cpu, -1)
+                cls._online(target, cpu, not up)
+                return False
+            except TargetStableError:
+                return True
+        except TargetStableError:
+            return False
+
+    @classmethod
+    def _prepare_hotplug(cls, target, cpu, up):
+        target.hotplug.fail(cpu, -1)
+        cls._online(target, cpu, not up)
+
+    @classmethod
+    def _get_states(cls, target, cpu, up):
+        states = target.hotplug.get_states()
+        cls._prepare_hotplug(target, cpu, not up)
+        return [
+                state
+                for state in states if cls._state_can_fail(target, cpu,
+                                                           state, up)
+        ]
+
+    @classmethod
+    def _mark_trace(cls, target, collector, start=0,
+                    expected=0, up=0, fail=0):
+        target.write_value(
+                collector.marker_file,
+                "HotplugRollback: test={} expected={} up={} fail={}".format(
+                    int(start), int(expected), int(up), fail),
+                verify=False
+        )
+
+    @classmethod
+    def _test_rollback(cls, target, collector, cpu, fail, up):
+        cls._prepare_hotplug(target, cpu, up=up)
+        target.hotplug.fail(cpu, fail)
+        cls._mark_trace(target, collector, start=1, up=up, fail=fail)
+        cls._online(target, cpu, online=up, verify=False)
+        cls._mark_trace(target, collector, start=0)
+
+    @classmethod
+    def _from_target(cls, target, *,
+                     res_dir: ArtifactPath = None) -> 'HotplugRollback':
+        target.hotplug.online_all()
+        cpu = target.hotplug.list_hotpluggable_cpus()[0]
+        collector = FtraceCollector(target, events=['cpuhp*'])
```

Comment:
The collector configuration will be provided by FtraceTestBundle, see RTATestBundle.run_rtapp() for an example